modules/oe_translation_poetry/tests/Functional/ContactFieldsWidgetTest.php

```diff
@@ -0,0 +1,193 @@
+<?php
+
+namespace Drupal\Tests\oe_translation_poetry\Functional;
+
+use Drupal\Core\Entity\Entity\EntityFormDisplay;
+use Drupal\field\Entity\FieldConfig;
+use Drupal\field\Entity\FieldStorageConfig;
+use Drupal\node\NodeInterface;
+use Drupal\Tests\BrowserTestBase;
+
+/**
+ * Tests the contact field widgets.
+ */
+class ContactFieldsWidgetTest extends BrowserTestBase {
+
+  /**
+   * The Node storage.
+   *
+   * @var \Drupal\node\NodeStorageInterface
+   */
+  protected $nodeStorage;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = [
+    'system',
+    'oe_translation_poetry',
+    'node',
+    'user',
+    'field',
+    'text',
+    'options',
+  ];
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp() {
+    parent::setUp();
+
+    $this->nodeStorage = \Drupal::entityTypeManager()->getStorage('node');
+
+    foreach ($this->getTestCases() as $case) {
+      FieldStorageConfig::create([
+        'field_name' => $case['field'],
+        'entity_type' => 'node',
+        'type' => $case['field_type'],
+      ])->save();
+
+      \Drupal::entityTypeManager()->getStorage('node_type')->create([
+        'name' => $case['bundle'],
+        'type' => $case['bundle'],
+      ])->save();
+
+      FieldConfig::create([
+        'entity_type' => 'node',
+        'field_name' => $case['field'],
+        'bundle' => $case['bundle'],
+      ])->save();
+
+      $entity_form_display = EntityFormDisplay::collectRenderDisplay($this->nodeStorage->create(['type' => $case['bundle']]), 'default');
+      $entity_form_display->setComponent($case['field'], [
+        'weight' => 1,
+        'region' => 'content',
+        'type' => $case['widget'],
+        'settings' => [],
+        'third_party_settings' => [],
+      ]);
+      $entity_form_display->save();
+    }
+
+    $user = $this->drupalCreateUser([], NULL, TRUE);
+    $this->drupalLogin($user);
+  }
+
+  /**
+   * Prepare the test cases for the two field types.
+   *
+   * @return array
+   *   The test cases.
+   */
+  protected function getTestCases(): array {
```

Comment:
This looks and behaves exactly like a data provider method, so why not make it a data provider?

Reply:
becasue we dont want a test run for each of the cases